### PR TITLE
Trigger deployment as success when new replicaSet has reached min you need

### PR DESCRIPTION
**Pull request by kiich**
Current check for deployment as success is to get related replicasets and then wait till all pods are ready.

This PR is to fine tune this logic where we get the new replica set associated to the deployment, wait until this new replicaset has reached minimum we require which is calculated from number of replicas needed in deployment minus value of maxUnavailable.

Once this value is reached, we treat the deployment is ready.

### CODE REVIEW

**Comment by kiich** [comment]
I signed it!

File: pkg/kube/client.go [hunk 1]
```diff
@@ -596,13 +615,32 @@ func (c *Client) waitForResources(timeout time.Duration, created Result) error {
 				if err != nil {
 					return false, err
 				}
+
 				for _, r := range rs.Items {
-					list, err := getPods(client, value.Namespace, r.Spec.Selector.MatchLabels)
+					currentReplicaSet, err := client.ReplicaSets(value.Namespace).Get(r.Name)
```
Comment (thomastaylor312) [inline pkg/kube/client.go]: Does this `Get` have to happen? They should be up to date from the Replica sets query above. You should just be able to just use `rs.Items`

File: pkg/kube/client.go [hunk 2]
```diff
@@ -596,13 +615,32 @@ func (c *Client) waitForResources(timeout time.Duration, created Result) error {
...
+				deployment, err := client.Deployments(value.Namespace).Get(value.Name)
+				if err != nil {
+					return false, err
+				}
+				// Find RS associated with deployment
+				newReplicaSet, err := deploymentutil.FindNewReplicaSet(deployment, replicaSets)
+				if err != nil {
+					return false, err
+				}
+				newDeployment := struct {
```
Comment (thomastaylor312) [inline pkg/kube/client.go]: Because you use this struct in 3 places, it might just be cleaner/easier to create it as a type

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
This is a smart logic update. Just a few changes and then I'll do some more testing with it

File: pkg/kube/client.go (hunk 2 above)
Comment (kiich) [inline pkg/kube/client.go]: Yep good point. Created as type in same file as I wasn't 100% sure where types used just in this file should go. Let me know if it needs to go to another place.

File: pkg/kube/client.go (hunk 1 above)
Comment (kiich) [inline pkg/kube/client.go]: Good catch. Amended.

**Comment by kiich** [comment]
Not entirely sure why circleci failed with the error on memory limit after my latest commit. @thomastaylor312 is there anything extra I can do from my side?

**Comment by thomastaylor312** [comment]
@kiich I am guessing that isn't a problem with your code. We have been hitting the limit with some CI things. I'll test this afternoon or over the weekend. Thanks for the quick response!

File: pkg/kube/client.go [hunk 3]
```diff
@@ -59,6 +61,12 @@ type Client struct {
 	SchemaCacheDir string
 }
 
+// Deployments hold associated replicaSets for deployments
+type Deployments struct {
```
Comment (thomastaylor312) [inline pkg/kube/client.go]: This should probably be unexported because other packages won't need to use it

**Comment by thomastaylor312** [comment]
@kiich I am done testing and things look good! I had one comment about this. It looks like it works as intended, but it does change the current behavior in one way. Right now as the `--wait` flag functions, it doesn't matter how many replicas you have, it just matters that all of the pods are ready. With this, if someone has `replicas` set to 1 and hasn't set `maxUnavailable` to 0 in their Update Strategy (1 is the default) or to `type: Recreate`, it will return as ready when it actually isn't. I know that is perfectly logical with how Deployments work, but I worry about other users being confused when the behavior changes on them. We can make it clear in the release notes, but if you have any other ideas on what we could do, I'd love to hear them.

Could you please update the documentation [here](https://github.com/kubernetes/helm/blob/15576abd9aaf7f285a7701905dbafbb256161c87/docs/using_helm.md) to reflect the change in behavior in 2.3 and also the `--help` documentation for the `rollback`, `upgrade`, and `install` commands? Once that is in, we should be good to merge. Sorry for the back and forth on this, I am really excited to see this land!

File: pkg/kube/client.go (hunk 3 above)
Comment (kiich) [inline pkg/kube/client.go]: Good point - fixing.

**Comment by kiich** [comment]
@thomastaylor312 Great to hear that! Good point on the `replicas` set to 1 and `maxUnavailable` set to/default to 1 case - the only thing I can think of right now is to catch when the "minAvailable" (deployment replicas - maxUnavailable basically) is 0 and set to 1 in `deploymentsReady` to ensure behaviour is consistent with other scenarios?

Do you want me to wait updating the docs until we figure out the above?

**Comment by kiich** [comment]
@thomastaylor312 Although thinking about it a bit more, by setting "minAvailable" to 1 in `deploymentsReady` when user didn't explicitly ask for it kind of feels wrong to me where as highlighting the change in behaviour in dcoumentation/release notes might suffice. (i.e. leaving the behaviour as per current PR)

Let me know your thoughts.

**Comment by kiich** [comment]
Also @thomastaylor312 when using `type: Recreate` I was able to observe that the new wait logic still waited until the new replicaset was ready instead of returning as ready immediately?

**Comment by thomastaylor312** [comment]
@kiich Sorry about that. I had written that and then tested again and forgot to remove it. `type: Recreate` works as it should. 

I think we'll just go with the documentation route on this one. Once that is pushed up, I'll read through it and then merge if it looks good

**Comment by thomastaylor312** [comment]
@kiich Just circling back on this one. We want to get this in to 2.3 and so are just waiting on that documentation before merging.

**Comment by kiich** [comment]
Hey @thomastaylor312 Sorry for the delay in this - was at KubeCon Europe
last week so hadn't had a chance to finish it off.

Will make it a priority!

Thanks

On Sat, 1 Apr 2017 at 20:26, Taylor Thomas <notifications@github.com> wrote:

@kiich <https://github.com/kiich> Just circling back on this one. We want
to get this in to 2.3 and so are just waiting on that documentation before
merging.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/helm/pull/2143#issuecomment-290941676>, or mute
the thread

**Comment by kiich** [comment]
@thomastaylor312 Apologies for the delay - Docs updated and hopefully what you were looking for. 

I had one question in that I wasn't sure how much detail to go into regards to the scenario where `replicas` is set to 1 and `maxUnavilable` not set to 0 - I assumed we could put most of the explanation in the release notes rather than in the `--help` and md doc?

**Comment by kiich** [comment]
Ah got it - further update to the md doc done. Let me know if that is enough/need further clarification.

Sorry about the extra changes in the man pages and cobra generated md docs - pretty new to this so was too keen to make sure everywhere was updated!

**Comment by thomastaylor312** [comment]
@kiich That documentation change looks great. I can give this a merge as soon as you rebase and if you can pop off the cobra and man doc changes. Thanks for all of the hard work!

**Comment by thomastaylor312** [comment]
Oh, and I'll figure out tests. It looks like the memory issue/spurious failure we have been running into lately

**Comment by kiich** [comment]
That's great! Commits squashed and doc changes removed. Let me know if anything needs doing!

**Comment by thomastaylor312** [comment]
@kiich This still needs a rebase against master to solve the merge conflict.

**Comment by kiich** [comment]
Ah apologies - I didn't get around in working on the conflict. I've used the github feature to resolve it as I was interested in trying that out - hope this is still ok.

**Review by thomastaylor312 — APPROVED** [review]

**Pull request merged**